### PR TITLE
Calculate BoolConstraint from ObjectConstraint in SE for equality operator

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Related to #5860 

Commit history shows what code changes updates what scenarios

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs [hunk 1]
```diff
@@ -28,21 +28,45 @@ internal static class Binary
...
+        {
+            if (left.HasConstraint<BoolConstraint>() && right.HasConstraint<BoolConstraint>())
+            {
+                return BinaryBoolConstraint(kind, left.HasConstraint(BoolConstraint.True), right.HasConstraint(BoolConstraint.True));
+            }
+            else if (left.HasConstraint<ObjectConstraint>() && right.HasConstraint<ObjectConstraint>())
+            {
+                return BinaryNullConstraint(kind, left.HasConstraint(ObjectConstraint.Null), right.HasConstraint(ObjectConstraint.Null));
+            }
+            else
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:45]: We should also support this matrix:
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/boolean-logical-operators#nullable-boolean-logical-operators

Are we following the lifting rules?
https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/nullable-value-types#lifted-operators

These could be follow-ups.

File: analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs [hunk 2]
```diff
@@ -291,5 +291,142 @@ public void Binary_UnexpectedOperator_VisitsBothBranches()
                 "Else",
                 "End");
         }
+
+
```
Comment (martin-strecker-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/SymbolicExecution/Roslyn/RoslynSymbolicExecutionTest.Binary.cs:295]:
```suggestion
```

**Review by martin-strecker-sonarsource — COMMENTED** [review]
LGTM. One extra newline and I think some special cases are missing.

**Review by martin-strecker-sonarsource — CHANGES_REQUESTED** [review]
I just want to make sure https://github.com/SonarSource/sonar-dotnet/pull/5918/files#r932044054 is addressed somehow.

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs [hunk 3]
```diff
@@ -28,21 +28,45 @@ internal static class Binary
...
+        private static SymbolicConstraint BinaryBoolConstraint(BinaryOperatorKind kind, bool left, bool right) =>
+                    kind switch
+                    {
+                        BinaryOperatorKind.Equals or BinaryOperatorKind.ObjectValueEquals => BoolConstraint.From(left == right),
+                        BinaryOperatorKind.NotEquals or BinaryOperatorKind.ObjectValueNotEquals => BoolConstraint.From(left != right),
+                        BinaryOperatorKind.And or BinaryOperatorKind.ConditionalAnd => BoolConstraint.From(left && right),
+                        BinaryOperatorKind.Or or BinaryOperatorKind.ConditionalOr => BoolConstraint.From(left || right),
+                        BinaryOperatorKind.ExclusiveOr => BoolConstraint.From(left ^ right),
+                        _ => null
+                    };
```
Comment (martin-strecker-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:60]: Intendation

File: analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:45]: The lifting will be most likely part of #4250 that is part of our sprint

Comment (martin-strecker-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/SymbolicExecution/Roslyn/OperationProcessors/Binary.cs:45]: Okay. I left a comment there to not lose the links to the C# spec.

**Review by martin-strecker-sonarsource — APPROVED** [review]

**Pull request merged**